inc/experimental/saidtel.h

```diff
@@ -0,0 +1,562 @@
+ *    not use this file except in compliance with the License. You may obtain
+ *    a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR
+ *    CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT
+ *    LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS
+ *    FOR A PARTICULAR PURPOSE, MERCHANTABILITY OR NON-INFRINGEMENT.
+ *
+ *    See the Apache Version 2.0 License for specific language governing
+ *    permissions and limitations under the License.
+ *
+ *    Microsoft would like to thank the following companies for their review and
+ *    assistance with these files: Intel Corporation, Mellanox Technologies Ltd,
+ *    Dell Products, L.P., Facebook, Inc
+ *
+ * @file    saidtel.h
+ *
+ * @brief       This module defines SAI data plane telemetry interface
+ * @description Supported by: Barefoot Networks, Inc.
+ * @warning     This module is a SAI experimental module.
+ */
+
+#if !defined (__SAIDTEL_H_)
+#define __SAIDTEL_H_
+
+#include <saitypes.h>
+
+/**
+ * @brief DTel attributes
+ *
+ * Only one DTel object per switch can be created.
+ */
+typedef enum _sai_dtel_attr_t
+{
+    /**
+     * @brief DTel INT endpoint
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default false
+     */
+    SAI_DTEL_ATTR_INT_ENDPOINT_ENABLE,
+
+    /**
+     * @brief DTel INT transit
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default false
+     */
+    SAI_DTEL_ATTR_INT_TRANSIT_ENABLE,
+
+    /**
+     * @brief Packet postcard
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default false
+     */
+    SAI_DTEL_ATTR_POSTCARD_ENABLE,
+
+    /**
+     * @brief Drop Report
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default false
+     */
+    SAI_DTEL_ATTR_DROP_REPORT_ENABLE,
+
+    /**
+     * @brief Queue Report
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default false
+     */
+    SAI_DTEL_ATTR_QUEUE_REPORT_ENABLE,
+
+    /**
+     * @brief Globally unique switch ID
+     *
+     * @type sai_uint32_t
+     * @flags CREATE_AND_SET
+     */
+    SAI_DTEL_ATTR_SWITCH_ID,
+
+    /**
+     * @brief DTel flow state clear cycle
+     *
+     * @type sai_uint16_t
+     * @flags CREATE_AND_SET
+     * @default 0
+     */
+    SAI_DTEL_ATTR_FLOW_STATE_CLEAR_CYCLE,
+
+    /**
+     * @brief Latency sensitivity for flow state change detection
+     *
+     * @type sai_uint8_t
+     * @flags CREATE_AND_SET
+     */
+    SAI_DTEL_ATTR_LATENCY_SENSITIVITY,
+
+    /**
+     * @brief DTel sink ports
+     *
+     * @type sai_object_list_t
+     * @flags CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_PORT
+     */
+    SAI_DTEL_ATTR_SINK_PORT_LIST,
+
+    /**
+     * @brief Reserved DSCP value for INT over L4
+     *
+     * @type sai_acl_field_data_t
+     * @flags CREATE_AND_SET
+     */
+    SAI_DTEL_ATTR_INT_L4_DSCP,
```

Comment:
For INT over L4, is this attribute mandatory?   
There is scenario where DSCP field is used for choosing packet queue.  If the original DSCP got changed,  packet may go through different queue which causes the report having wrong egress timestamp and queue info.

Reply:
It is not mandatory. One current design choice is to use DSCP field to identify INT information after L4 header. Only one DSCP bit is needed to be reserved for INT, while other DSCP bits can still be used for choosing packet queue. People can choose to use other fields before L4 header to differentiate INT packets, which requires a new attribute in SAI_DTEL.

Reply:
Since DSCP is a 6-bit field, taking one DSCP bit means 32 dscp values are gone.  I think we'll need new attribute in SAI_DTEL to support other INT design choices like the probe marker solution mentioned in https://tools.ietf.org/html/draft-lapukhov-dataplane-probe-01

Reply:
So there is no standard on the INT packet transportation format yet, but sink and transit nodes need mechanism to identify them.  Could some attribute with opaque value be defined besides SAI_DTEL_ATTR_INT_L4_DSCP?   The opaque value is up to ASIC SDK to interpret then act accordingly?  SAI just provides the configuration channel.

Reply:
That's a nice idea. I suggest us to give more thoughts and discussions on whether it is better to add an opaque value now, or add another specific field later when someone plans to use INT with some other specific format. One great thing about SAI is that it is not RFC. It is more flexible for the users/community to add more fields as the network applications evolve.